runtime/signals.c

```diff
@@ -29,101 +28,144 @@
-       in caml_garbage_collection
+   - via total_signals_are_pending, processed in
+     caml_process_pending_signals_exn.
+   - by playing with the allocation limit, processed in
+     caml_garbage_collection
 */
 
+CAMLno_tsan
 CAMLexport void caml_record_signal(int signal_number)
 {
-  caml_pending_signals[signal_number] = 1;
-  caml_signals_are_pending = 1;
+  atomic_fetch_add_explicit(&caml_pending_signals[signal_number], 1, memory_order_relaxed);
+  atomic_fetch_add_explicit(&total_signals_pending, 1, memory_order_release);
   caml_interrupt_self();
 }
 
-int caml_init_signal_stack()
+/* Management of blocking sections. */
+
+static void caml_enter_blocking_section_default(void)
 {
-#ifdef POSIX_SIGNALS
-  stack_t stk;
-  stk.ss_flags = 0;
-  stk.ss_size = SIGSTKSZ;
-  stk.ss_sp = caml_stat_alloc_noexc(stk.ss_size);
-  if(stk.ss_sp == NULL) {
-    return -1;
-  }
-  if (sigaltstack(&stk, NULL) < 0) {
-    caml_stat_free(stk.ss_sp);
-    return -1;
-  }
+  caml_bt_exit_ocaml();
+  caml_release_domain_lock();
+}
 
-  /* gprof installs a signal handler for SIGPROF.
-     Make it run on the alternate signal stack, to prevent segfaults. */
-  {
-    struct sigaction act;
-    sigaction(SIGPROF, NULL, &act);
-    if ((act.sa_flags & SA_SIGINFO) ||
-        (act.sa_handler != SIG_IGN && act.sa_handler != SIG_DFL)) {
-      /* found a handler */
-      if ((act.sa_flags & SA_ONSTACK) == 0) {
-        act.sa_flags |= SA_ONSTACK;
-        sigaction(SIGPROF, &act, NULL);
-      }
-    }
-  }
-#endif
-  return 0;
+static void caml_leave_blocking_section_default(void)
+{
+  caml_bt_enter_ocaml();
+  caml_acquire_domain_lock();
 }
 
-void caml_free_signal_stack()
+CAMLexport void (*caml_enter_blocking_section_hook)(void) =
+   caml_enter_blocking_section_default;
+CAMLexport void (*caml_leave_blocking_section_hook)(void) =
+   caml_leave_blocking_section_default;
+
+CAMLexport void caml_enter_blocking_section(void)
 {
-#ifdef POSIX_SIGNALS
-  stack_t stk, disable = {0};
-  disable.ss_flags = SS_DISABLE;
-  /* POSIX says ss_size is ignored when SS_DISABLE is set,
-     but OSX/Darwin fails if the size isn't set. */
-  disable.ss_size = SIGSTKSZ;
-  if (sigaltstack(&disable, &stk) < 0) {
-    caml_fatal_error_arg("Failed to reset signal stack: %s", strerror(errno));
+  while (1){
+    /* Process all pending signals now */
+    caml_raise_if_exception(caml_process_pending_signals_exn());
+    caml_enter_blocking_section_hook ();
+    /* Check again for pending signals.
+       If none, done; otherwise, try again */
+    if ( atomic_load_explicit(&total_signals_pending, memory_order_relaxed) == 0 ) break;
+    caml_leave_blocking_section_hook ();
   }
-  caml_stat_free(stk.ss_sp);
-#endif
 }
 
-/* Execute a signal handler immediately */
+CAMLexport void caml_enter_blocking_section_no_pending(void)
+{
+  caml_enter_blocking_section_hook ();
+}
+
+CAMLexport void caml_leave_blocking_section(void)
```

Comment:
I'm not sure if there is a deliberate change here. Previously we had:
```C
CAMLexport void caml_leave_blocking_section() {
  caml_leave_blocking_section_hook();
  caml_process_pending_signals();
}
```
I see upstream having:
```C
CAMLexport void caml_leave_blocking_section(void)
{
  int saved_errno;
  /* Save the value of errno (PR#5982). */
  saved_errno = errno;
  caml_leave_blocking_section_hook ();

   ...
  if (check_for_pending_signals()) {
    signals_are_pending = 1;
    caml_set_action_pending();
  }

  errno = saved_errno;
}
```
Is it deliberate to lose the processing of pending signals in `caml_leave_blocking_section` or the setting of a flag (or other)?

Reply:
It's worth pointing out that the one you have from trunk (which the EINTR pretty much matches other than check_for_pending_signals) doesn't actually process pending signals but just checks if they're pending due because it's possible for `signals_are_pending` to be flagged off even if there is one available in the specific signal flags.

We don't have that problem in this PR because we're using counters.